File: leetcode/382-linked-list-random-node.scala
// https://leetcode.com/problems/linked-list-random-node

/**
  * class ListNode(
  *   var x: Int = 0,
  *   var next: ListNode = null
  * )
  */

class Solution(head: ListNode) {

  import scala.util.Random

  def getRandom(): Int = {
    def randomOneNth(n: Int): Boolean = (Random nextInt n) == 0
    def h(n: Int, node: ListNode, z: Int): Int =
      if (node eq null) z
      else h(n + 1, node.next, if (randomOneNth(n)) node.x else z)
    h(1, head, 0)
  }

}

File: leetcode/382-linked-list-random-node.py
# https://leetcode.com/problems/linked-list-random-node

# pylint: disable=I0011, C0103, C0111

import random

# Definition for singly-linked list.
# class ListNode:
#   def __init__(self, x):
#     self.val = x
#     self.next = None

class Solution:

  def __init__(self, head):
    """
    @param head The linked list's head.
    Note that the head is guaranteed to be not null, so it contains at least one node.
    :type head: ListNode
    """
    self.head = head

  def getRandom(self):
    """
    Returns a random node's value.
    :rtype: int
    """
    def random_one_nth(n):
      return random.randint(1, n) == 1
    i = 1
    n = self.head
    v = self.head.val
    while n.next:
      i += 1
      n = n.next
      if random_one_nth(i):
        v = n.val
    return v
